render-ws-java-client/src/main/java/org/janelia/render/client/MultiSemFlatFieldCorrectionClient.java

```diff
@@ -0,0 +1,195 @@
+				final ResolvedTileSpecCollection tileSpecs = renderClient.getResolvedTiles(stack.getStack(), (double) z);
+
+				for (final TileSpec tileSpec : tileSpecs.getTileSpecs()) {
+					final ImageProcessor ip = loadImageTile(tileSpec);
+					final int sfov = extractSfovNumber(tileSpec);
+					final ImageProcessor flatFieldEstimate = loadFlatFieldEstimate(Math.min(z, params.flatFieldConstantFromZ), sfov);
+
+					applyFlatFieldCorrection(ip, flatFieldEstimate);
+
+					patchTileSpec(tileSpec);
+					saveImage(ip, tileSpec);
+				}
+
+				renderClient.saveResolvedTiles(tileSpecs, stack.getStack() + params.targetStackSuffix, (double) z);
+			}
+			renderClient.setStackState(stack.getStack() + params.targetStackSuffix, StackMetaData.StackState.COMPLETE);
+		}
+	}
+
+	private DoubleSummaryStatistics getZStatistics(final StackId stack) throws IOException {
+		final List<Double> zValues = renderClient.getStackZValues(stack.getStack());
+		final DoubleSummaryStatistics zStatistics = zValues.stream().mapToDouble(Double::doubleValue).summaryStatistics();
+		LOG.info("Considering {} with {} z-layers", stack, zStatistics.getCount());
+		return zStatistics;
+	}
+
+	private ImageProcessor loadFlatFieldEstimate(final int z, final int sfov) {
+		final Path imagePath = Path.of(params.flatFieldLocation, String.format(params.flatFieldFormat, z, sfov));
+		final String imageUrl = "file:" + imagePath;
+		final ImageLoader.LoaderType loaderType = ImageLoader.LoaderType.IMAGEJ_DEFAULT;
+		return CACHE.get(imageUrl, 0, false, false, loaderType, null);
+	}
+
+	private int extractSfovNumber(final TileSpec tileSpec) {
+		final String tileId = tileSpec.getTileId();
+		final String[] parts = tileId.split("_");
+		return Integer.parseInt(parts[1].substring(4));
+	}
+
+	/**
+	 * Apply the flat field estimate to the input image.
+	 * @param ip the input image that is altered in place
+	 * @param flatFieldEstimate the flat field estimate to apply
+	 */
+	private void applyFlatFieldCorrection(final ImageProcessor ip, final ImageProcessor flatFieldEstimate) {
+		// convert to 32-bit grayscale (float) for lossless processing
+		final FloatProcessor fp = ip.convertToFloatProcessor();
+
+		for (int i = 0; i < ip.getPixelCount(); i++) {
+			final double a = fp.getf(i);
+			final double b = flatFieldEstimate.getf(i);
+			fp.setf(i, (float) (a / b));
+		}
+
+		// convert back to original bit depth
+		fp.setMinAndMax(0, 255);
+		ip.setPixels(0, fp);
+	}
+
+	private void patchTileSpec(final TileSpec tileSpec) {
+		final Path originalPath = Path.of(tileSpec.getTileImageUrl().replaceFirst("file:", ""));
+		final Path newPath;
+
+		if (params.inputRoot != null) {
+			final Path relativePath = Path.of(params.inputRoot).relativize(originalPath);
+			newPath = Path.of(params.outputFolder).resolve(relativePath);
+		} else {
+			newPath = Path.of(params.outputFolder).resolve(originalPath.getFileName());
+		}
+
+		ensureFolderExists(newPath.getParent());
```

Comment:
I would separate file system stuff from tile spec manipulation and ensure the folder exists as part of the saveImage process.